Mathlib/Algebra/Lie/Weights/RootSystem.lean

```diff
@@ -0,0 +1,248 @@
+  rw [add_comm, chainBotCoeff_add_chainTopCoeff]
+
+lemma chainLength_sub_chainTopCoeff :
+    chainLength α β - chainTopCoeff α β = chainBotCoeff α β :=
+  Nat.sub_eq_of_eq_add (chainBotCoeff_add_chainTopCoeff α β).symm
+
+lemma chainLength_sub_chainBotCoeff :
+    chainLength α β - chainBotCoeff α β = chainTopCoeff α β :=
+  Nat.sub_eq_of_eq_add (chainTopCoeff_add_chainBotCoeff α β).symm
+
+lemma chainBotCoeff_le_chainLength : chainBotCoeff α β ≤ chainLength α β :=
+  (Nat.le_add_left _ _).trans_eq (chainTopCoeff_add_chainBotCoeff α β)
+
+/-- If `β - qα ... β ... β + rα` is the `α`-chain through `β`, then
+  `β (coroot α) = q - r`. In particular, it is an integer. -/
+lemma apply_coroot :
+    β (coroot α) = (chainBotCoeff α β - chainTopCoeff α β : ℤ) := by
+  rw [apply_coroot_eq_cast, ← chainTopCoeff_add_chainBotCoeff]; congr 1; omega
+
+/-- Members of the `α`-chain through `β` are the only roots of the form `β - kα`. -/
+lemma rootSpace_zsmul_add_ne_bot_iff (n : ℤ) :
+    rootSpace H (n • α + β) ≠ ⊥ ↔ n ≤ chainTopCoeff α β ∧ -n ≤ chainBotCoeff α β := by
+  constructor
+  · contrapose
+    rw [not_and_or, not_le, not_le, ne_eq, not_not]
+    rintro (h | h)
+    · lift n to ℕ using (Nat.cast_nonneg _).trans h.le
+      rw [Nat.cast_lt, ← @Nat.add_lt_add_iff_left (chainBotCoeff α β),
+        chainBotCoeff_add_chainTopCoeff, ← chainLength_neg] at h
+      have := rootSpace_neg_nsmul_add_chainTop_of_lt (-α) β hα.neg h
+      rw [coe_chainTop, LieModule.Weight.coe_neg, smul_neg, smul_neg, neg_neg, ← neg_smul,
+        ← add_assoc, nsmul_eq_smul_cast ℤ, ← add_smul, chainTopCoeff_neg, Nat.cast_add,
+        ← sub_eq_add_neg, add_sub_cancel_left] at this
+      exact this
+    · obtain ⟨n, rfl⟩ := neg_involutive.surjective n
+      simp only [neg_neg] at h
+      lift n to ℕ using (Nat.cast_nonneg _).trans h.le
+      rw [Nat.cast_lt, ← @Nat.add_lt_add_iff_right (chainTopCoeff α β),
+        chainBotCoeff_add_chainTopCoeff] at h
+      have := rootSpace_neg_nsmul_add_chainTop_of_lt α β hα h
+      rwa [nsmul_eq_smul_cast ℤ, ← neg_smul, coe_chainTop, ← add_assoc,
+        ← add_smul, Nat.cast_add, neg_add, add_assoc, neg_add_self, add_zero] at this
+  · rintro ⟨h₁, h₂⟩
+    set k := chainTopCoeff α β - n with hk; clear_value k
+    lift k to ℕ using (by rw [hk, le_sub_iff_add_le, zero_add]; exact h₁)
+    rw [eq_sub_iff_add_eq, ← eq_sub_iff_add_eq'] at hk
+    subst hk
+    simp only [neg_sub, tsub_le_iff_right, ← Nat.cast_add, Nat.cast_le,
+      chainBotCoeff_add_chainTopCoeff] at h₂
+    have := rootSpace_neg_nsmul_add_chainTop_of_le α β h₂
+    rwa [coe_chainTop, nsmul_eq_smul_cast ℤ, ← neg_smul,
+      ← add_assoc, ← add_smul, ← sub_eq_neg_add] at this
+
+lemma rootSpace_zsmul_add_ne_bot_iff_mem (n : ℤ) :
+    rootSpace H (n • α + β) ≠ ⊥ ↔ n ∈ Finset.Icc (-chainBotCoeff α β : ℤ) (chainTopCoeff α β) := by
+  rw [rootSpace_zsmul_add_ne_bot_iff α β hα n, Finset.mem_Icc, and_comm, neg_le]
```

Comment:
Again I wouldn't bother adding this unless it was repeatedly useful
```suggestion
```

Reply:
I think this is useful to keep. When we are able to calculate `q = chainBotCoeff α β` and `r = chainTopCoeff α β`, then the finset is defeq to `{-q, -q + 1, ..., r-1, r}`.